File: ios/App/Plugins/Phrase/PipelineController.swift
import Foundation

/// Orquestrador principal do pipeline de detecção no iOS.
/// Gerencia state machine, integra 4 layers, aplica segurança. Paridade com PipelineController.java.
class PipelineController {
    /// Protocolo para ações do pipeline
    protocol ActionListener: AnyObject {
        func onPipelineAction(action: String, phraseId: String?, confidence: Double, reasons: [String])
    }

    private let config: ScoringConfig
    private let enrollmentManager: PhraseEnrollmentManager
    private let phraseDetector: PhraseDetector
    private let riskScorer: RiskScorer
    private let repetitionTracker: CommandRepetitionTracker
    private let ringBuffer: RingAudioBuffer
    private let metrics = PipelineMetrics()

    private(set) var currentState: PipelineState = .idleMonitoring
    weak var actionListener: ActionListener?

    init(config: ScoringConfig, enrollmentManager: PhraseEnrollmentManager) {
        self.config = config
        self.enrollmentManager = enrollmentManager
        self.phraseDetector = PhraseDetector(config: config)
        self.riskScorer = RiskScorer(config: config)
        self.repetitionTracker = CommandRepetitionTracker(config: config)
        self.ringBuffer = RingAudioBuffer(durationSeconds: config.confirmationWindowPreSeconds + config.confirmationWindowPostSeconds)

        print("[PipelineController-iOS] Inicializado | estado=\(currentState.rawValue)")
    }

    // MARK: - Entradas do pipeline (chamados pelo AudioTriggerNativePlugin)

    /// Alimenta áudio bruto no ring buffer (Layer 2)
    func feedAudio(samples: [Int16], offset: Int, length: Int) {
        ringBuffer.write(samples: samples, offset: offset, length: length)
        metrics.l1FramesProcessed += 1
    }

    /// Alimenta RMS para detecção de padrão estável
    func feedRmsDb(_ rmsDb: Double) {
        riskScorer.feedRmsDb(rmsDb)
    }

    /// Trigger acústico do detector existente (Layer 1 → Layer 3)
    func onAcousticTrigger(speechDensity: Double, reason: String) {
        metrics.l1AcousticTriggers += 1
        metrics.lastTriggerTime = Date()

        guard !enrollmentManager.loadedTemplates.isEmpty else { return }
        guard currentState == .idleMonitoring || currentState == .commandListening else { return }

        // Transição de estado
        currentState = .confirmationWindow
        metrics.l2BufferSnapshots += 1

        // Capturar snapshot do buffer
        let audioSnapshot = ringBuffer.snapshotAll()
        guard audioSnapshot.count > 0 else {
            currentState = .idleMonitoring
            return
        }

        // Layer 3: Análise Mode A (triggered)
        currentState = .phraseAnalysis
        metrics.l3AnalysesModeA += 1

        let detections = phraseDetector.analyzeModeA(
            audioSamples: audioSnapshot,
            templates: enrollmentManager.loadedTemplates)

        if !detections.isEmpty {
            metrics.l3MatchesModeA += detections.count
            metrics.lastMatchTime = Date()

            // Layer 4: Risk scoring
            currentState = .riskScoring
            metrics.l4ScoreCalculations += 1

            let scoringResult = riskScorer.calculateScore(
                speechDensity: speechDensity,
                loudDensity: 0,
                rmsDb: 0,
                detectedPhrases: detections)

            // Processar comandos operacionais
            processOperationalCommands(detections: detections)

            // Notificar ação se score alto
            if scoringResult.level == .high || scoringResult.level == .critical {
                metrics.l4ActionsTriggered += 1
                metrics.lastActionTime = Date()
                currentState = .actionTriggered

                actionListener?.onPipelineAction(
                    action: "RISK_\(scoringResult.level.rawValue)",
                    phraseId: detections.first?.phraseId,
                    confidence: scoringResult.score,
                    reasons: scoringResult.reasons)
            }
        }

        // Voltar ao idle
        currentState = .idleMonitoring
    }

    // MARK: - Comandos operacionais

    private func processOperationalCommands(detections: [PhraseDetector.DetectionResult]) {
        for detection in detections where detection.type == .operational {
            let commandId = detection.phraseId

            // Verificar segurança do comando
            guard let security = config.commandSecurity[commandId] else { continue }

            // Verificar cooldown
            if repetitionTracker.isInCooldown(commandId: commandId, cooldownSeconds: security.cooldownSeconds) {
                print("[PipelineController-iOS] Comando '\(commandId)' em cooldown — ignorado")
                continue
            }

            // Verificar confidence mínima
            guard detection.confidence >= security.confidenceMin else {
                print("[PipelineController-iOS] Comando '\(commandId)' conf=\(String(format: "%.2f", detection.confidence)) < \(security.confidenceMin) — ignorado")
                continue
            }

            // Registrar repetição
            repetitionTracker.recordMatch(commandId: commandId, confidence: detection.confidence)

            // Verificar repetições necessárias
            let repetitions = repetitionTracker.getRecentRepetitions(commandId: commandId)
            if repetitions >= security.repetitionsRequired {
                // Executar ação
                repetitionTracker.recordAction(commandId: commandId)
                metrics.l4ActionsTriggered += 1
                metrics.lastActionTime = Date()

                print("[PipelineController-iOS] Comando '\(commandId)' executado | conf=\(String(format: "%.2f", detection.confidence)) | reps=\(repetitions)")

                actionListener?.onPipelineAction(
                    action: "COMMAND",
                    phraseId: commandId,
                    confidence: detection.confidence,
                    reasons: ["Comando operacional reconhecido"])
            }
        }
    }

    // MARK: - Diagnósticos

    func getDiagnostics() -> String {
        return "[Pipeline-iOS] state=\(currentState.rawValue) | \(metrics.summary())"
    }

    func reset() {
        currentState = .idleMonitoring
        riskScorer.reset()
        repetitionTracker.reset()
        ringBuffer.reset()
        metrics.reset()
    }
}

File: android/app/src/main/java/tech/orizon/ampara/audio/phrase/PipelineController.java
package tech.orizon.ampara.audio.phrase;

import android.content.Context;
import android.content.Intent;
import android.util.Log;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Controlador do pipeline de detecção multiestágio (SPEC seções 5, 6, 7).
 * 
 * Orquestra os 4 layers, gerencia a state machine, aplica regras de segurança
 * por comando, e mantém métricas de observabilidade.
 * 
 * Fluxo principal:
 * L1 (existente) → trigger → L2 (buffer) → L3 (phrase detection) → L4 (scoring)
 * → ação
 */
public class PipelineController {
    private static final String TAG = "PipelineCtrl";

    // Dependências
    private final Context context;
    private final RingAudioBuffer ringBuffer;
    private final PhraseDetector phraseDetector;
    private final RiskScorer riskScorer;
    private final ScoringConfig config;
    private final PipelineMetrics metrics;
    private final PhraseEnrollmentManager enrollmentManager;

    // Estado
    private PipelineState currentState = PipelineState.IDLE_MONITORING;
    private boolean modeBEnabled = false;

    // Trackers de repetição por comando
    private final Map<String, CommandRepetitionTracker> commandTrackers;

    // Post-trigger: controla a coleta de áudio pós-trigger
    private boolean collectingPostTrigger = false;
    private long postTriggerStartTime;
    private short[] preTriggerAudio;

    // Callback para ações
    private PipelineActionListener actionListener;

    /**
     * Interface para comunicar ações do pipeline ao AudioTriggerService
     */
    public interface PipelineActionListener {
        void onStartRecording(String reason);

        void onStopRecording(String reason);

        void onStartPanic(String reason);

        void onCancelPanic(String reason);

        void onElevateSensitivity(String reason);

        void onRiskScoreUpdate(double score, String level, List<String> reasons);

        void onPhraseDetected(String phraseId, double confidence, String type);
    }

    public PipelineController(Context context, ScoringConfig config,
            PhraseEnrollmentManager enrollmentManager) {
        this.context = context;
        this.config = config;
        this.enrollmentManager = enrollmentManager;

        this.ringBuffer = new RingAudioBuffer(
                config.preTriggerSeconds + config.postTriggerSeconds, 16000);
        this.phraseDetector = new PhraseDetector(enrollmentManager, config);
        this.riskScorer = new RiskScorer(config);
        this.metrics = new PipelineMetrics();

        // Criar trackers para comandos operacionais
        this.commandTrackers = new HashMap<>();
        commandTrackers.put("start_recording", new CommandRepetitionTracker(
                "start_recording", config.repetitionHistoryMaxEntries, config.repetitionHistoryMaxAgeMs));
        commandTrackers.put("stop_recording", new CommandRepetitionTracker(
                "stop_recording", config.repetitionHistoryMaxEntries, config.repetitionHistoryMaxAgeMs));
        commandTrackers.put("start_panic", new CommandRepetitionTracker(
                "start_panic", config.repetitionHistoryMaxEntries, config.repetitionHistoryMaxAgeMs));
        commandTrackers.put("cancel_panic", new CommandRepetitionTracker(
                "cancel_panic", config.repetitionHistoryMaxEntries, config.repetitionHistoryMaxAgeMs));

        Log.i(TAG, "[INIT] PipelineController criado | estado=IDLE_MONITORING");
    }

    /**
     * Registra listener de ações
     */
    public void setActionListener(PipelineActionListener listener) {
        this.actionListener = listener;
    }

    // ========== Alimentação contínua de áudio ==========

    /**
     * Chamado continuamente pelo AudioTriggerService a cada frame de áudio.
     * Alimenta o ring buffer e, se em post-trigger, coleta a janela.
     *
     * @param samples áudio PCM
     * @param offset  início
     * @param length  tamanho
     */
    public void feedAudio(short[] samples, int offset, int length) {
        // Alimentar ring buffer continuamente
        ringBuffer.write(samples, offset, length);

        // Se estamos coletando pós-trigger
        if (collectingPostTrigger) {
            long elapsed = System.currentTimeMillis() - postTriggerStartTime;
            if (elapsed >= config.postTriggerSeconds * 1000L) {
                // Post-trigger completo: montar janela e analisar
                collectingPostTrigger = false;
                onConfirmationWindowReady();
            }
        }
    }

    /**
     * Alimenta RMS em dB para o detector de padrão estável
     */
    public void feedRmsDb(double rmsDb) {
        riskScorer.updateRmsHistory(rmsDb);
    }

    // ========== Trigger de L1 (evento acústico) ==========

    /**
     * Chamado quando o detector acústico existente (L1) gera um trigger.
     * Inicia o processo de confirmação.
     *
     * @param acousticSuspicionScore score acústico de L1 (0.0-1.0)
     * @param eventType              tipo de evento ("SPEECH_BURST", "IMPACT", etc.)
     */
    public void onAcousticTrigger(double acousticSuspicionScore, String eventType) {
        metrics.recordL1Activation();

        // Verificar se deve escalar
        if (acousticSuspicionScore < 0.3 && !"IMPACT".equals(eventType)) {
            return; // L1 ativou mas score baixo demais para escalar
        }

        metrics.recordL1Escalation();
        Log.i(TAG, String.format("[TRIGGER] L1 escalou | score=%.2f | event=%s | estado=%s",
                acousticSuspicionScore, eventType, currentState));

        // Capturar pre-trigger do buffer
        preTriggerAudio = ringBuffer.snapshot(config.preTriggerSeconds);
        if (preTriggerAudio == null) {
            Log.w(TAG, "[TRIGGER] Buffer insuficiente para snapshot");
            return;
        }

        // Iniciar coleta pós-trigger
        collectingPostTrigger = true;
        postTriggerStartTime = System.currentTimeMillis();
        currentState = PipelineState.CONFIRMATION_WINDOW;

        Log.d(TAG, String.format("[TRIGGER] Pre-trigger=%d amostras | aguardando %ds pós-trigger",
                preTriggerAudio.length, config.postTriggerSeconds));
    }

    /**
     * Chamado quando a confirmation window está completa (pre + post trigger)
     */
    private void onConfirmationWindowReady() {
        currentState = PipelineState.PHRASE_ANALYSIS;

        // Montar janela completa: snapshot do buffer (que agora inclui o áudio
        // pós-trigger)
        short[] fullWindow = ringBuffer.snapshot(
                config.preTriggerSeconds + config.postTriggerSeconds);

        if (fullWindow == null) {
            Log.w(TAG, "[WINDOW] Janela de confirmação vazia");
            currentState = PipelineState.IDLE_MONITORING;
            return;
        }

        metrics.recordL2Window(fullWindow.length * 2);

        Log.d(TAG, String.format("[WINDOW] Janela pronta: %d amostras (%.1fs)",
                fullWindow.length, (double) fullWindow.length / 16000));

        // L3: Análise de frases (Modo A — triggered completo)
        PhraseDetector.DetectionResult detectionResult = phraseDetector.analyzeTriggered(fullWindow, 0,
                fullWindow.length);

        metrics.recordL3Result(detectionResult);

        // Processar comandos operacionais ANTES do scoring (SPEC seção 6.2)
        boolean commandExecuted = processOperationalCommands(detectionResult);

        if (!commandExecuted) {
            // L4: Scoring
            currentState = PipelineState.RISK_SCORING;
            // Usar score acústico estimado baseado na existência do trigger
            RiskScorer.RiskResult riskResult = riskScorer.calculateRisk(
                    0.5, "SPEECH_BURST", detectionResult, 0);

            metrics.recordL4Score(riskResult.score, riskResult.triggeredAction != null);

            // Notificar listener
            if (actionListener != null) {
                actionListener.onRiskScoreUpdate(
                        riskResult.score, riskResult.level, riskResult.reasons);

                if (riskResult.triggeredAction != null) {
                    executeAction(riskResult.triggeredAction, riskResult.reasons.toString());
                }
            }
        }

        // Voltar ao idle
        currentState = PipelineState.IDLE_MONITORING;
    }

    // ========== Modo B — Escuta contínua ==========

    /**
     * Chamado periodicamente (a cada ~3s) quando Modo B está ativo.
     * Analisa uma janela curta buscando apenas comandos operacionais.
     *
     * @param samples áudio PCM dos últimos 3s
     * @param offset  início
     * @param length  tamanho
     */
    public void analyzeModeBWindow(short[] samples, int offset, int length) {
        if (!modeBEnabled)
            return;

        metrics.recordModeBWindow();

        PhraseDetector.DetectionResult result = phraseDetector.analyzeContinuous(samples, offset, length);

        if (result.hasMatches()) {
            processOperationalCommands(result);
        }
    }

    // ========== Processamento de comandos operacionais ==========

    /**
     * Processa matches operacionais aplicando a matriz de segurança (SPEC seção 6).
     * Vocabulário genérico NUNCA aciona comandos (regra formal).
     *
     * @return true se algum comando foi executado
     */
    private boolean processOperationalCommands(PhraseDetector.DetectionResult result) {
        List<PhraseDetector.PhraseMatch> opMatches = result.getOperationalMatches();
        if (opMatches.isEmpty())
            return false;

        for (PhraseDetector.PhraseMatch match : opMatches) {
            // Vocabulário genérico NUNCA dispara ação
            if (match.type == PhraseTemplate.PhraseType.GENERIC)
                continue;

            CommandRepetitionTracker tracker = commandTrackers.get(match.phraseId);
            if (tracker == null)
                continue;

            // Registrar match no tracker
            tracker.recordMatch(match.confidence, match.dtwDistance);

            // Verificar cooldown
            int cooldown = config.getCooldownMs(match.phraseId);
            if (!tracker.isCooldownExpired(cooldown)) {
                Log.d(TAG, String.format("[COMMAND] %s em cooldown, ignorando", match.phraseId));
                continue;
            }

            // Aplicar regras de segurança por tipo de comando
            if (isCommandAuthorized(match, tracker)) {
                tracker.markActionExecuted();
                executeAction(match.phraseId, String.format("conf=%.2f", match.confidence));

                // Notificar phrase detected
                if (actionListener != null) {
                    actionListener.onPhraseDetected(
                            match.phraseId, match.confidence, match.type.name());
                }
                return true;
            }
        }
        return false;
    }

    /**
     * Aplica a matriz de segurança por tipo de comando (SPEC seção 6.1)
     */
    private boolean isCommandAuthorized(PhraseDetector.PhraseMatch match,
            CommandRepetitionTracker tracker) {
        switch (match.phraseId) {
            case "start_recording":
                // 🟢 Permissivo: confidence >= 0.70
                return match.confidence >= config.startRecordingConfidence;

            case "stop_recording":
                // 🟡 Moderado: confidence >= 0.80
                return match.confidence >= config.stopRecordingConfidence;

            case "start_panic":
                // 🔴 Restritivo: confidence >= 0.92 (instantâneo)
                // OU confidence >= 0.85 + repetição 2x em 15s
                if (match.confidence >= config.startPanicInstantConfidence) {
                    return true;
                }
                return match.confidence >= config.startPanicConfidence
                        && tracker.hasRepetitions(2, config.startPanicRepetitionWindowMs);

            case "cancel_panic":
                // 🔴🔴 Ultra-restritivo: SEMPRE exige confidence >= 0.90 + 2x em 20s
                return match.confidence >= config.cancelPanicConfidence
                        && tracker.hasRepetitions(
                                config.cancelPanicRepetitions,
                                config.cancelPanicRepetitionWindowMs);

            default:
                return match.confidence >= 0.75;
        }
    }

    /**
     * Executa uma ação operacional ou baseada em score
     */
    private void executeAction(String action, String reason) {
        if (actionListener == null)
            return;

        Log.i(TAG, String.format("[ACTION] %s | reason=%s", action, reason));

        switch (action) {
            case "start_recording":
            case "START_RECORDING":
                actionListener.onStartRecording(reason);
                break;
            case "stop_recording":
                actionListener.onStopRecording(reason);
                break;
            case "start_panic":
            case "START_PANIC":
                actionListener.onStartPanic(reason);
                break;
            case "cancel_panic":
                actionListener.onCancelPanic(reason);
                break;
            case "ELEVATE_SENSITIVITY":
                actionListener.onElevateSensitivity(reason);
                break;
        }
    }

    // ========== Controle de estado ==========

    public void setModeBEnabled(boolean enabled) {
        this.modeBEnabled = enabled;
        Log.i(TAG, "[MODE] Modo B (continuous) = " + enabled);
    }

    public boolean isModeBEnabled() {
        return modeBEnabled;
    }

    public PipelineState getCurrentState() {
        return currentState;
    }

    public PipelineMetrics getMetrics() {
        return metrics;
    }

    /**
     * Pausa o pipeline (ex: interrupção de áudio no iOS)
     */
    public void pause(String reason) {
        Log.i(TAG, String.format("[PAUSE] Pipeline pausado | reason=%s | estado=%s",
                reason, currentState));
        collectingPostTrigger = false;
    }

    /**
     * Retoma o pipeline após pausa
     */
    public void resume() {
        Log.i(TAG, "[RESUME] Pipeline retomado");
        currentState = PipelineState.IDLE_MONITORING;
    }

    /**
     * Registra um gap de monitoramento (iOS interrupção definitiva)
     */
    public void logGap(long startedAt) {
        long duration = System.currentTimeMillis() - startedAt;
        Log.w(TAG, String.format("[GAP] Gap de monitoramento: %.1fs", duration / 1000.0));
    }

    /**
     * Reseta todo o estado do pipeline
     */
    public void reset() {
        currentState = PipelineState.IDLE_MONITORING;
        collectingPostTrigger = false;
        ringBuffer.reset();
        riskScorer.reset();
        metrics.reset();
        for (CommandRepetitionTracker tracker : commandTrackers.values()) {
            tracker.reset();
        }
        Log.i(TAG, "[RESET] Pipeline resetado");
    }

    /**
     * Retorna informações de diagnóstico para o painel técnico
     */
    public String getDiagnostics() {
        return String.format("State=%s | ModeB=%b | %s | %s",
                currentState, modeBEnabled,
                metrics.toDisplayString(),
                phraseDetector.getDiagnostics());
    }
}
